### PR TITLE
[FIX] website_sale: multi-company payment methods 

### PR DESCRIPTION
**Pull request by yosa-odoo**
Steps to reproduce:
- Have a portal user with a related contact (thus on the contact, partner_share is True)
- Set a default company A on the user, and the same company A on the sales tab on the contact (so the company_id field is A)
- Have a website linked to company B
- Connect to said website with the portal user
- Try to buy something

Issue:
Internal error - Multi company issue

Solution:
It is mainly a configuration error but the message is not user firendly.
Instead we make sure the partner cn proceed a transaction in the company website

opw-2734190

### CODE REVIEW

**Comment by yosa-odoo** [comment]
if approved, I'll amend the `.pot` file

**Comment by yosa-odoo** [comment]
Method used in https://github.com/odoo/odoo/pull/98910

**Review by agr-odoo — COMMENTED** [review]
After reading https://github.com/odoo/odoo/pull/98910 it seems `_can_partner_pay_in_company` is used to detect cases where we need `sudo`

Here we would use it to filter after retrieving acquirers from DB, which in general is not ideal.
We could detect that no acquires suit our needs before searching them 
```diff
diff --git a/addons/website_sale/controllers/main.py b/addons/website_sale/controllers/main.py
index 20b636663f00..2b4a504e5af5 100644
--- a/addons/website_sale/controllers/main.py
+++ b/addons/website_sale/controllers/main.py
@@ -888,17 +888,20 @@ class WebsiteSale(http.Controller):
             bootstrap_formatting= True
         )
 
+        if order.partner_id.company_id and order.partner_id.company_id.id != order.company_id.id:
+            values['errors'].append(
+                (_('Sorry, we are unable to find a Payment Method'),
+                 _('No payment method is available for your current order. '
+                   'Please contact us for more information.')))
+            return values
+
         domain = expression.AND([
```

**Comment by yosa-odoo** [comment]
@agr-odoo Makes sense. I was going for another solution but it can be as straightforward than what you propose :)
Thanks!

**Comment by agr-odoo** [comment]
Should be ok, we need the `.pot` and commit msg updated

**Review by agr-odoo — APPROVED** [review]
@AntoineVDV Could you have at look at this? thanks

**Comment by AntoineVDV** [comment]
@Williambraecky Your codeowner was broken. It should now be fixed for all PRs targeting website_sale or documents, sorry! :grin:

**Comment by AntoineVDV** [comment]
> After reading #98910 it seems `_can_partner_pay_in_company` is used to detect cases where we need `sudo`

This is actually done in 15.0+ because it is impossible to reconcile the `account.payment` created from the transaction with the partner if they have a different company.

This PR's fix LGTM but I'll let William handle the in-depth review since it's an eCommerce-only change.

File: addons/website_sale/controllers/main.py
Comment (Williambraecky) [inline addons/website_sale/controllers/main.py:890]: This might be too restrictive if we were to someday relax the security rules on payment providers.
But I guess it's unlikely to happen and I am ok with it as is right now.

**Review by Williambraecky — APPROVED** [review]
Thanks for the fix.

robodoo r+

**Pull request closed without merging**
